
<h1> DS200A Computer Vision Assignment</h1>

Author: Junsheng Pei
        Nuocheng Lyu
        Zihao Yang

<h2>  Part Three: Training Models </h2>	

In [1]:
import sklearn.model_selection as model_selection
import numpy as np
import matplotlib.pyplot as plt
from pathlib import Path
import pandas as pd
import matplotlib
import skimage

from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification
from sklearn import svm

In [2]:
data_file = Path("./", "feature_matrix.hdf")
feature_matrix_train = pd.read_hdf(data_file, "train")
feature_matrix_test = pd.read_hdf(data_file, "test")

In [3]:
def train_test_split(df):  
    train_label = df['label']
    df = df.drop(columns = ['label'])
    X_train, X_val, y_train, y_val = model_selection.train_test_split(df,
                train_label, test_size=0.2, random_state=0)
        
    return X_train ,X_val, y_train, y_val

def accuracy(pred, actual):
    return sum(pred == actual) / len(pred)
# Calculate the accuracy percentage of the predicted values

<h3>  Train models using all of the following methods below. Be sure to drop the actual image column, and the encoding</h3>	Take note of the differences in accuracy, and methods.


In [4]:
X_train ,X_val, y_train, y_val = train_test_split(feature_matrix_train)

Logistic Regression

In [5]:
clf_LR = LogisticRegression(fit_intercept=True)
clf_LR.fit(X_train, y_train)

y_pred_train = clf_LR.predict(X_train)
y_pred_val = clf_LR.predict(X_val)

print('training accuracy:  ' + str(accuracy(y_pred_train,y_train)))
print('test accuracy:  ' + str(accuracy(y_pred_val,y_val)))

training accuracy:  0.23916666666666667
test accuracy:  0.22259136212624583


/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:459: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


K-nearest Neighbors

In [6]:
for i in [1,2,5,7,10,15,20,25,30]:
    clf_KNN = KNeighborsClassifier(n_neighbors = i)
    clf_KNN.fit(X_train, y_train)

    y_pred_train = clf_KNN.predict(X_train)
    y_pred_val = clf_KNN.predict(X_val)
    print('when n_neighbors is {0:.4f}: training accuracy: {1:.4f} \
          validation accuracy: {2:.4f}'.format(i , accuracy(y_pred_train,y_train), accuracy(y_pred_val,y_val)))

when n_neighbors is 1.0000: training accuracy: 1.0000           validation accuracy: 0.2193
when n_neighbors is 2.0000: training accuracy: 0.6017           validation accuracy: 0.1927
when n_neighbors is 5.0000: training accuracy: 0.4100           validation accuracy: 0.1927
when n_neighbors is 7.0000: training accuracy: 0.3642           validation accuracy: 0.1927
when n_neighbors is 10.0000: training accuracy: 0.3292           validation accuracy: 0.1960
when n_neighbors is 15.0000: training accuracy: 0.2775           validation accuracy: 0.1894
when n_neighbors is 20.0000: training accuracy: 0.2817           validation accuracy: 0.1794
when n_neighbors is 25.0000: training accuracy: 0.2692           validation accuracy: 0.1595
when n_neighbors is 30.0000: training accuracy: 0.2608           validation accuracy: 0.1628


Random Forest

In [7]:
clf_forest = RandomForestClassifier(n_estimators=200, max_depth=13, random_state=5)
clf_forest.fit(X_train, y_train)
print('training accuracy:  ' + str(clf_forest.score(X_train, y_train)))
print('test accuracy:  ' + str(clf_forest.score(X_val, y_val)))

training accuracy:  0.9983333333333333
test accuracy:  0.4318936877076412


Support Vector Machine

In [8]:
#clf = svm.SVC(kernel='linear',C = 1)
clf_svm = svm.SVC(decision_function_shape = 'ovo', random_state = 0, max_iter = 10000)
clf_svm.fit(X_train, y_train)
print('test accuracy:  ' + str(clf_svm.score(X_train,y_train)))
print('test accuracy:  ' + str((clf_svm.score(X_val,y_val))))

test accuracy:  1.0
test accuracy:  0.12624584717607973


/anaconda3/lib/python3.6/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


## With the the Random Forest, we can get the best validation accuracy 43.2%

 ### Export the results of test data with Random Forest

In [9]:
y_test = clf_forest.predict(feature_matrix_test)

In [10]:
# save the numpy array into cvs without the last comma
with open('test_results.csv', 'w') as file:
    for x in y_test[:-2]:
        file.write('{0},'.format(x))
    file.write(str(y_test[-1]) )
file.close()    